File: pkg/v1/features/crons.go
// Deprecated: This package is part of the legacy v0 workflow definition system.
// Use the new Go SDK at github.com/hatchet-dev/hatchet/sdks/go instead. Migration guide: https://docs.hatchet.run/home/migration-guide-go
package features

import (
	"context"

	"github.com/google/uuid"
	"github.com/robfig/cron/v3"

	"github.com/hatchet-dev/hatchet/pkg/client/rest"
)

// Deprecated: CronsClient is part of the old generics-based v1 Go SDK.
// Use the new Go SDK at github.com/hatchet-dev/hatchet/sdks/go instead. Migration guide: https://docs.hatchet.run/home/migration-guide-go
type CronsClient interface {
	Create(ctx context.Context, workflowName string, cron CreateCronTrigger) (*rest.CronWorkflows, error)

	Delete(ctx context.Context, cronId string) error

	List(ctx context.Context, opts rest.CronWorkflowListParams) (*rest.CronWorkflowsList, error)

	Get(ctx context.Context, cronId string) (*rest.CronWorkflows, error)
}

// Deprecated: CreateCronTrigger is part of the old generics-based v1 Go SDK.
// Use the new Go SDK at github.com/hatchet-dev/hatchet/sdks/go instead. Migration guide: https://docs.hatchet.run/home/migration-guide-go
type CreateCronTrigger struct {
	Name               string                 `json:"name"`
	Expression         string                 `json:"expression"`
	Input              map[string]interface{} `json:"input,omitempty"`
	AdditionalMetadata map[string]interface{} `json:"additionalMetadata,omitempty"`
	Priority           *int32                 `json:"priority,omitempty"`
}

// cronsClientImpl implements the CronsClient interface.
type cronsClientImpl struct {
	api      *rest.ClientWithResponses
	tenantId uuid.UUID
}

// Deprecated: NewCronsClient is part of the old generics-based v1 Go SDK.
// Use the new Go SDK at github.com/hatchet-dev/hatchet/sdks/go instead. Migration guide: https://docs.hatchet.run/home/migration-guide-go
func NewCronsClient(
	api *rest.ClientWithResponses,
	tenantId *string,
) CronsClient {
	tenantIdUUID := uuid.MustParse(*tenantId)

	return &cronsClientImpl{
		api:      api,
		tenantId: tenantIdUUID,
	}
}

// Deprecated: ValidateCronExpression is part of the old generics-based v1 Go SDK.
// Use the new Go SDK at github.com/hatchet-dev/hatchet/sdks/go instead. Migration guide: https://docs.hatchet.run/home/migration-guide-go
func ValidateCronExpression(expression string) bool {
	parser := cron.NewParser(cron.Minute | cron.Hour | cron.Dom | cron.Month | cron.Dow)
	_, err := parser.Parse(expression)

	return err == nil
}

// Deprecated: Create is part of the old generics-based v1 Go SDK.
// Use the new Go SDK at github.com/hatchet-dev/hatchet/sdks/go instead. Migration guide: https://docs.hatchet.run/home/migration-guide-go
//
// Create creates a new cron workflow trigger.
func (c *cronsClientImpl) Create(ctx context.Context, workflowName string, cron CreateCronTrigger) (*rest.CronWorkflows, error) {
	// Validate cron expression
	if !ValidateCronExpression(cron.Expression) {
		return nil, &InvalidCronExpressionError{Expression: cron.Expression}
	}

	// Prepare input and metadata maps if nil
	input := cron.Input
	if input == nil {
		input = make(map[string]interface{})
	}

	additionalMetadata := cron.AdditionalMetadata
	if additionalMetadata == nil {
		additionalMetadata = make(map[string]interface{})
	}

	request := rest.CronWorkflowTriggerCreateJSONRequestBody{
		CronName:           cron.Name,
		CronExpression:     cron.Expression,
		Input:              input,
		AdditionalMetadata: additionalMetadata,
	}

	resp, err := c.api.CronWorkflowTriggerCreateWithResponse(
		ctx,
		c.tenantId,
		workflowName,
		request,
	)
	if err != nil {
		return nil, err
	}

	return resp.JSON200, nil
}

// Deprecated: Delete is part of the old generics-based v1 Go SDK.
// Use the new Go SDK at github.com/hatchet-dev/hatchet/sdks/go instead. Migration guide: https://docs.hatchet.run/home/migration-guide-go
//
// Delete removes a cron workflow trigger.
func (c *cronsClientImpl) Delete(ctx context.Context, cronId string) error {
	cronIdUUID, err := uuid.Parse(cronId)
	if err != nil {
		return err
	}

	_, err = c.api.WorkflowCronDeleteWithResponse(
		ctx,
		c.tenantId,
		cronIdUUID,
	)
	return err
}

// Deprecated: List is part of the old generics-based v1 Go SDK.
// Use the new Go SDK at github.com/hatchet-dev/hatchet/sdks/go instead. Migration guide: https://docs.hatchet.run/home/migration-guide-go
//
// List retrieves a collection of cron workflow triggers based on the provided parameters.
func (c *cronsClientImpl) List(ctx context.Context, opts rest.CronWorkflowListParams) (*rest.CronWorkflowsList, error) {
	resp, err := c.api.CronWorkflowListWithResponse(
		ctx,
		c.tenantId,
		&opts,
	)
	if err != nil {
		return nil, err
	}

	return resp.JSON200, nil
}

// Deprecated: Get is part of the old generics-based v1 Go SDK.
// Use the new Go SDK at github.com/hatchet-dev/hatchet/sdks/go instead. Migration guide: https://docs.hatchet.run/home/migration-guide-go
//
// Get retrieves a specific cron workflow trigger by its ID.
func (c *cronsClientImpl) Get(ctx context.Context, cronId string) (*rest.CronWorkflows, error) {
	cronIdUUID, err := uuid.Parse(cronId)
	if err != nil {
		return nil, err
	}

	resp, err := c.api.WorkflowCronGetWithResponse(
		ctx,
		c.tenantId,
		cronIdUUID,
	)
	if err != nil {
		return nil, err
	}

	return resp.JSON200, nil
}

// Deprecated: InvalidCronExpressionError is part of the old generics-based v1 Go SDK.
// Use the new Go SDK at github.com/hatchet-dev/hatchet/sdks/go instead. Migration guide: https://docs.hatchet.run/home/migration-guide-go
type InvalidCronExpressionError struct {
	Expression string
}

// Deprecated: Error is part of the old generics-based v1 Go SDK.
// Use the new Go SDK at github.com/hatchet-dev/hatchet/sdks/go instead. Migration guide: https://docs.hatchet.run/home/migration-guide-go
func (e *InvalidCronExpressionError) Error() string {
	return "invalid cron expression: " + e.Expression
}

File: sdks/typescript/src/v1/client/features/crons.ts
import { CronWorkflows, CronWorkflowsList } from '@hatchet/clients/rest/generated/data-contracts';
import { z } from 'zod';
import { Workflow } from '@hatchet/workflow';
import { AxiosError } from 'axios';
import { isValidUUID } from '@util/uuid';
import { BaseWorkflowDeclaration } from '@hatchet/v1';
import { applyNamespace } from '@hatchet/util/apply-namespace';
import { HatchetClient } from '../client';
import { workflowNameString, WorkflowsClient } from './workflows';

/**
 * Validates a cron expression to support both 5-field and 6-field formats.
 * Also supports timezone prefixes like CRON_TZ= and TZ=.
 * @param expression - The cron expression to validate.
 * @returns True if valid, false otherwise.
 */
function validateCronExpression(expression: string): boolean {
  if (!expression || typeof expression !== 'string') {
    return false;
  }

  // Extract cron part for field count validation (timezone is supported but we validate the cron part)
  let cronPart = expression;
  if (expression.startsWith('CRON_TZ=')) {
    const parts = expression.split(' ', 2);
    if (parts.length === 2) {
      cronPart = parts[1];
    }
  } else if (expression.startsWith('TZ=')) {
    const parts = expression.split(' ', 2);
    if (parts.length === 2) {
      cronPart = parts[1];
    }
  }

  const fields = cronPart.trim().split(/\s+/);

  // Only allow 5 or 6 fields (not 7 with year field)
  return fields.length === 5 || fields.length === 6;
}

/**
 * Schema for creating a Cron Trigger.
 */
export const CreateCronTriggerSchema = z.object({
  name: z.string(),
  expression: z.string().refine(validateCronExpression, {
    message: 'Cron expression must have 5 fields (minute hour day month weekday) or 6 fields (second minute hour day month weekday)',
  }),
  input: z.record(z.any()).optional(),
  additionalMetadata: z.record(z.string()).optional(),
  priority: z.number().optional(),
});

/**
 * Type representing the input for creating a Cron.
 */
export type CreateCronInput = z.infer<typeof CreateCronTriggerSchema>;

/**
 * Client for managing Cron Triggers.
 */
export class CronClient {
  api: HatchetClient['api'];
  tenantId: string;
  workflows: WorkflowsClient;
  namespace: string | undefined;

  constructor(client: HatchetClient) {
    this.api = client.api;
    this.tenantId = client.tenantId;
    this.workflows = new WorkflowsClient(client);
    this.namespace = client.config.namespace;
  }

  /**
   * Retrieves the Cron ID from a CronWorkflows object or a string.
   * @param cron - The CronWorkflows object or Cron ID as a string.
   * @returns The Cron ID as a string.
   */
  private getCronId(cron: CronWorkflows | string): string {
    const str = typeof cron === 'string' ? cron : cron.metadata.id;

    if (!isValidUUID(str)) {
      throw new Error('Invalid cron ID: must be a valid UUID');
    }

    return str;
  }

  /**
   * Creates a new Cron workflow.
   * @param workflow - The workflow identifier or Workflow object.
   * @param cron - The input data for creating the Cron Trigger.
   * @returns A promise that resolves to the created CronWorkflows object.
   * @throws Will throw an error if the input is invalid or the API call fails.
   */
  async create(
    workflow: string | Workflow | BaseWorkflowDeclaration<any, any>,
    cron: CreateCronInput
  ): Promise<CronWorkflows> {
    const workflowId = applyNamespace(workflowNameString(workflow), this.namespace);

    // Validate cron input with zod schema
    try {
      const parsedCron = CreateCronTriggerSchema.parse(cron);
      const response = await this.api.cronWorkflowTriggerCreate(this.tenantId, workflowId, {
        cronName: parsedCron.name,
        cronExpression: parsedCron.expression,
        input: parsedCron.input ?? {},
        additionalMetadata: parsedCron.additionalMetadata ?? {},
        priority: parsedCron.priority,
      });
      return response.data;
    } catch (err) {
      if (err instanceof z.ZodError) {
        throw new Error(`Invalid cron input: ${err.message}`);
      }

      if (err instanceof AxiosError) {
        throw new Error(JSON.stringify(err.response?.data.errors));
      }

      throw err;
    }
  }

  /**
   * Deletes an existing Cron Trigger.
   * @param cron - The Cron Trigger ID as a string or CronWorkflows object.
   * @returns A promise that resolves when the Cron Trigger is deleted.
   */
  async delete(cron: string | CronWorkflows): Promise<void> {
    const cronId = this.getCronId(cron);
    await this.api.workflowCronDelete(this.tenantId, cronId);
  }

  /**
   * Lists all Cron Triggers based on the provided query parameters.
   * @param query - Query parameters for listing Cron Triggers.
   * @returns A promise that resolves to a CronWorkflowsList object.
   */
  async list(
    query: Parameters<typeof this.api.cronWorkflowList>[1] & {
      workflow?: string | Workflow | BaseWorkflowDeclaration<any, any>;
    }
  ): Promise<CronWorkflowsList> {
    const { workflow, ...rest } = query;

    if (workflow) {
      const workflowId = await this.workflows.getWorkflowIdFromName(
        applyNamespace(workflowNameString(workflow), this.namespace)
      );
      rest.workflowId = workflowId;
    }

    const response = await this.api.cronWorkflowList(this.tenantId, rest);
    return response.data;
  }

  /**
   * Retrieves a specific Cron Trigger by its ID.
   * @param cron - The Cron Trigger ID as a string or CronWorkflows object.
   * @returns A promise that resolves to the CronWorkflows object.
   */
  async get(cron: string | CronWorkflows): Promise<CronWorkflows> {
    const cronId = this.getCronId(cron);
    const response = await this.api.workflowCronGet(this.tenantId, cronId);
    return response.data;
  }
}
